app/Main.hs

```diff
@@ -20,26 +24,47 @@ main :: IO ()
 main = do
   opt <- parseOptions
   progName <- T.pack <$> getProgName
+  let logLevel  = opt ^. optLogLevel
+  let kafkaConf = opt ^. optKafkaConfig
+  let statsConf = opt ^. optStatsConfig
 
   withStdOutTimedFastLogger $ \logger -> do
-    env <- mkEnv (opt ^. optRegion) (opt ^. optLogLevel) logger
-    void . runApplication progName env opt logger $ do
-      logInfo "Instantiating Schema Registry"
-      sr <- schemaRegistry (opt ^. optKafkaSchemaRegistryAddress)
-
-      logInfo "Creating Kafka Consumer"
-      consumer <- mkConsumer opt
-      -- producer <- mkProducer opt -- Use this if you also want a producer.
-
-      logInfo "Running Kafka Consumer"
-      runConduit $
-        kafkaSourceNoClose consumer (opt ^. optKafkaPollTimeoutMs)
-        .| throwLeftSatisfy isFatal                   -- throw any fatal error
-        .| skipNonFatalExcept [isPollTimeout]         -- discard any non-fatal except poll timeouts
-        .| Srv.handleStream sr                        -- handle messages (see Service.hs)
-        -- .| batchByOrFlushEither (opt ^. optBatchSize) -- Use this if you also want a producer.
-        .| everyNSeconds (opt ^. optKafkaConsumerCommitPeriodSec)  -- only commit ever N seconds, so we don't hammer Kafka.
-        .| commitOffsetsSink consumer
-        -- .| flushThenCommitSink consumer producer -- Swap with the above if you want a producer.
+    withStatsClient progName statsConf $ \stats -> do
+      envAws <- mkEnv (opt ^. optRegion) logLevel logger
+      let envApp = AppEnv opt stats logger
+
+      void . runApplication envAws envApp $ do
+        logInfo "Creating Kafka Consumer"
+        consumer <- mkConsumer logLevel kafkaConf
+        -- producer <- mkProducer kafkaConf -- Use this if you also want a producer.
+
+        logInfo "Instantiating Schema Registry"
+        sr <- schemaRegistry (kafkaConf ^. schemaRegistryAddress)
+
+        logInfo "Running Kafka Consumer"
+        runConduit $
+          kafkaSourceNoClose consumer (kafkaConf ^. pollTimeoutMs)
+          .| throwLeftSatisfy isFatal                   -- throw any fatal error
+          .| skipNonFatalExcept [isPollTimeout]         -- discard any non-fatal except poll timeouts
+          .| tapRight (Srv.handleStream sr)             -- handle messages (see Service.hs)
+          -- .| batchByOrFlushEither (kafkaConf ^. batchSize) -- Use this if you also want a producer.
```

Comment:
Have you made sure that this still works?

Reply:
I have made sure that it won't work and I stated it in the PR description

Reply:
That's the point: it depends. In different cases you need different solutions.
One of them would be to batch before tapping into `handleStream` so it receives batches in the first place.
Another would be to update `handleStream` to be a `Conduit` and batch them there.
Third solution would be to batch it is a sink, remember offsets and to commit explicit offsets.
We use different strategies in different projects.